File: moqt/internal/message/setup_test.go
package message_test

import (
	"bytes"
	"testing"

	"github.com/qumo-dev/gomoqt/moqt/internal/message"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestSetupMessage_EncodeDecode(t *testing.T) {
	tests := map[string]struct {
		input message.SetupMessage
	}{
		"empty parameter list": {
			input: message.SetupMessage{},
		},
		"probe parameter": {
			input: func() message.SetupMessage {
				var sm message.SetupMessage
				sm.AddProbe(message.ProbeLevelReport)
				return sm
			}(),
		},
		"path parameter": {
			input: func() message.SetupMessage {
				var sm message.SetupMessage
				sm.AddPath("/relay/live")
				return sm
			}(),
		},
		"probe and path parameters": {
			input: func() message.SetupMessage {
				var sm message.SetupMessage
				sm.AddProbe(message.ProbeLevelIncrease)
				sm.AddPath("/")
				return sm
			}(),
		},
		"unknown parameter is preserved": {
			input: message.SetupMessage{
				Parameters: []message.SetupParameter{
					{ID: 0x7f, Value: []byte{0x01, 0x02}},
				},
			},
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var buf bytes.Buffer

			require.NoError(t, tt.input.Encode(&buf))

			var decoded message.SetupMessage
			require.NoError(t, decoded.Decode(&buf))

			assert.Equal(t, tt.input.ProbeLevel(), decoded.ProbeLevel())
			gotPath, gotOk := decoded.Path()
			wantPath, wantOk := tt.input.Path()
			assert.Equal(t, wantOk, gotOk)
			assert.Equal(t, wantPath, gotPath)
			assert.Len(t, decoded.Parameters, len(tt.input.Parameters))
		})
	}
}

func TestSetupMessage_ProbeLevel(t *testing.T) {
	tests := map[string]struct {
		input    message.SetupMessage
		expected uint64
	}{
		"absent means none": {
			input:    message.SetupMessage{},
			expected: message.ProbeLevelNone,
		},
		"report": {
			input: func() message.SetupMessage {
				var sm message.SetupMessage
				sm.AddProbe(message.ProbeLevelReport)
				return sm
			}(),
			expected: message.ProbeLevelReport,
		},
		"increase": {
			input: func() message.SetupMessage {
				var sm message.SetupMessage
				sm.AddProbe(message.ProbeLevelIncrease)
				return sm
			}(),
			expected: message.ProbeLevelIncrease,
		},
		"malformed value means none": {
			input: message.SetupMessage{
				Parameters: []message.SetupParameter{
					{ID: message.SetupParamProbe, Value: nil},
				},
			},
			expected: message.ProbeLevelNone,
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			assert.Equal(t, tt.expected, tt.input.ProbeLevel())
		})
	}
}

func TestSetupMessage_Decode_DuplicateParameter(t *testing.T) {
	var sm message.SetupMessage
	sm.AddProbe(message.ProbeLevelReport)
	sm.AddProbe(message.ProbeLevelIncrease)

	var buf bytes.Buffer
	require.NoError(t, sm.Encode(&buf))

	var decoded message.SetupMessage
	err := decoded.Decode(&buf)
	assert.ErrorIs(t, err, message.ErrDuplicateSetupParameter)
}

func TestSetupMessage_DecodeErrors(t *testing.T) {
	tests := map[string]struct {
		data []byte
	}{
		"empty input":         {data: []byte{}},
		"truncated parameter": {data: []byte{0x03, 0x01, 0x01, 0x05}}, // claims 5-byte value, none follow
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			var decoded message.SetupMessage
			err := decoded.Decode(bytes.NewReader(tt.data))
			assert.Error(t, err)
		})
	}
}

File: moq-web/src/internal/message/setup_test.ts
import { assertEquals } from "@std/assert";
import { ProbeLevels, SetupMessage } from "./setup.ts";
import { Buffer } from "@okdaichi/golikejs/bytes";

Deno.test("SetupMessage - encode/decode roundtrip", async (t) => {
	await t.step("empty parameter list", async () => {
		const buffer = Buffer.make(64);
		const message = new SetupMessage({});
		assertEquals(await message.encode(buffer), undefined);

		const decoded = new SetupMessage({});
		assertEquals(await decoded.decode(buffer), undefined);
		assertEquals(decoded.parameters.length, 0);
		assertEquals(decoded.probeLevel(), ProbeLevels.None);
		assertEquals(decoded.path(), undefined);
	});

	await t.step("probe parameter", async () => {
		const buffer = Buffer.make(64);
		const message = new SetupMessage({});
		message.addProbe(ProbeLevels.Report);
		assertEquals(await message.encode(buffer), undefined);

		const decoded = new SetupMessage({});
		assertEquals(await decoded.decode(buffer), undefined);
		assertEquals(decoded.probeLevel(), ProbeLevels.Report);
	});

	await t.step("path parameter", async () => {
		const buffer = Buffer.make(64);
		const message = new SetupMessage({});
		message.addPath("/relay/live");
		assertEquals(await message.encode(buffer), undefined);

		const decoded = new SetupMessage({});
		assertEquals(await decoded.decode(buffer), undefined);
		assertEquals(decoded.path(), "/relay/live");
	});

	await t.step("unknown parameter is preserved", async () => {
		const buffer = Buffer.make(64);
		const message = new SetupMessage({
			parameters: [{ id: 0x7f, value: new Uint8Array([1, 2]) }],
		});
		assertEquals(await message.encode(buffer), undefined);

		const decoded = new SetupMessage({});
		assertEquals(await decoded.decode(buffer), undefined);
		assertEquals(decoded.parameters.length, 1);
		assertEquals(decoded.parameters[0]?.id, 0x7f);
		assertEquals(decoded.parameters[0]?.value, new Uint8Array([1, 2]));
	});

	await t.step("duplicate parameter is rejected", async () => {
		const buffer = Buffer.make(64);
		const message = new SetupMessage({});
		message.addProbe(ProbeLevels.Report);
		message.addProbe(ProbeLevels.Increase);
		assertEquals(await message.encode(buffer), undefined);

		const decoded = new SetupMessage({});
		const err = await decoded.decode(buffer);
		assertEquals(err instanceof Error, true);
	});

	await t.step("decode returns error on empty input", async () => {
		const buffer = Buffer.make(0);
		const decoded = new SetupMessage({});
		const err = await decoded.decode(buffer);
		assertEquals(err !== undefined, true);
	});
});
